app/models/git_repository.rb

```diff
@@ -9,6 +9,11 @@ class GitRepository
     Rails.application.config.samson.cached_repos_dir
   end
 
+  def self.valid_ref_format?(reference)
+    IO.popen(['git', 'check-ref-format', "refs/head/#{reference}"]) { |_| }
```

Comment:
it might make more sense to just rescue the error... there's no rule saying a valid git reference will be parsed into a valid URL.
`check-ref-format 'refs/heads/master&asdf#111'; echo $?` printed `0` for me but I suspect anything after the `#` might not even get sent to the server, depending on your choice of HTTP library.